### DISCUSSION TITLE
Difference between @query and createRef

### DISCUSSION THREAD

**Discussion by dsokur** [lit/lit#4856] [Q&A] [answered]
It is not clear to me when to use @query decorator and when to use refs. Both provide you access to DOM element. What is more performant considering @query has a cache option? Basically what should I use @query and when should I use refs.

**Comment by sorvell** [comment, marked as answer]
They can often be used interchangeably, but here are some guidelines that might be helpful. 

The `@query` decorator creates a property on an element. It's best used when the maintaining the reference is important for the element or (in rare cases) the value should be part of the element's public API. And obviously you need a to be rendering a custom element in order to use `@query`. The value retrieved via `@query` can also be cached, which is a slight performance win if the value doesn't ever need to update.

The `ref` directive and `createRef` can be used with or without a custom element. It's best used when the ref is really a pure templating concern. A ref can be passed into a shared template function for example: (inputRef) => html`<input ${ref(inputRef)}>`. You can also provide a function to createRef which runs every time the ref updates and is useful for cases when the ref will change and/or work needs to be done when it does.

**Reply by dsokur** [reply]
@sorvell thanks
